Implement #11 by adding support for related models

Colin-Seifer opened the pull request:
implements feature #11 by adding support for related models to documents, indices, registries, and signals

## Additions in this PR
- Restores related models functionality by adding content to documents.py, indices.py, registries.py, and signals.py
- Adds fixtures back into the test suite
- Adds many tests to boost coverage

## Changes in this PR
- Updates documentation in document.md to include support for related_models

qcoumes commented on django_opensearch_dsl/registries.py:
```diff
@@ -79,6 +85,42 @@ def register_document(self, document):
+                    yield doc
+
+    def update_related(self, instance, **kwargs):
+        """Update docs that have related_models."""
```
Can you update to something in the line of :

```
"""Update documents related to `instance`.

Related documents are found using the `get_instances_from_related()` method of each Document classes including the instance's Model is including within their `related_models`.
"""
```

qcoumes commented on django_opensearch_dsl/registries.py:
```diff
@@ -79,6 +85,42 @@ def register_document(self, document):
+                doc_instance.update(related, **kwargs)
+
+    def delete_related(self, instance, **kwargs):
+        """Remove `instance` from related models."""
```
Can you detail the interaction with `related_instance_to_ignore` within the docstring ?

Colin-Seifer replied:
Added quite a bit to this docstring. Let me know if there needs to be better clarification on that somewhere.

qcoumes commented on tests/tests/test_indices.py:
```diff
@@ -0,0 +1,45 @@
+        index = Index("test")
+        self.assertEqual(index.__str__(), "test")
+
+    def test__init__(self):
+        settings.OPENSEARCH_DSL_INDEX_SETTINGS = {
+            "number_of_replicas": 0,
+            "number_of_shards": 2,
+        }
+
+        index = Index("test")
+        self.assertEqual(
+            index._settings,
+            {
+                "number_of_replicas": 0,
+                "number_of_shards": 2,
+            },
+        )
+
+        settings.OPENSEARCH_DSL_INDEX_SETTINGS = {}
```
Please use Django's `django.test.override_settings` either as a decorator or context manager instead of modifying `settings` directly.

qcoumes commented on tests/tests/test_registries.py:
```diff
@@ -0,0 +1,181 @@
+        doc_d1.get_instances_from_related.assert_not_called()
+        doc_d1.update.assert_not_called()
+
+    def test_autosync(self):
+        settings.OPENSEARCH_DSL_AUTOSYNC = False
+
+        instance = self.ModelA()
+        self.registry.update(instance)
+        self.assertFalse(self.doc_a1.update.called)
+
+        settings.OPENSEARCH_DSL_AUTOSYNC = True
```
Please use Django's `django.test.override_settings` either as a decorator or context manager instead of modifying `settings` directly.

qcoumes reviewed (changes requested):
Hi Colin,

Thanks for this great PR and the thorough testing!

I requested some changes, mostly about documentation and docstring, I can't wait for this feature to be available once again.

Do you think you can rebase/squash your commits so that only one (or two if you want to separate the tests in another commit) are fast-forwarded onto `master`?

Colin-Seifer commented:
> Hi Colin,
> 
> Thanks for this great PR and the thorough testing!
> 
> I requested some changes, mostly about documentation and docstring, I can't wait for this feature to be available once again.
> 
> Do you think you can rebase/squash your commits so that only one (or two if you want to separate the tests in another commit) are fast-forwarded onto `master`?

Hi Quentin,

Thanks for the comments! I'll get started on addressing all of these on Monday, and yes, I can go ahead and rebase/squash the commits.

Ameer-Amr commented:
> > Hi Colin,
> > Thanks for this great PR and the thorough testing!
> > I requested some changes, mostly about documentation and docstring, I can't wait for this feature to be available once again.
> > Do you think you can rebase/squash your commits so that only one (or two if you want to separate the tests in another commit) are fast-forwarded onto `master`?
> 
> Hi Quentin,
> 
> Thanks for the comments! I'll get started on addressing all of these on Monday, and yes, I can go ahead and rebase/squash the commits.

Hey Colin,
Is this PR ready to merge with Master? if not when we can expect this feature?

Colin-Seifer commented:
> > > Hi Colin,
> > > Thanks for this great PR and the thorough testing!
> > > I requested some changes, mostly about documentation and docstring, I can't wait for this feature to be available once again.
> > > Do you think you can rebase/squash your commits so that only one (or two if you want to separate the tests in another commit) are fast-forwarded onto `master`?
> > 
> > 
> > Hi Quentin,
> > Thanks for the comments! I'll get started on addressing all of these on Monday, and yes, I can go ahead and rebase/squash the commits.
> 
> Hey Colin, Is this PR ready to merge with Master? if not when we can expect this feature?

Not yet, I should finish addressing the comments by the end of today.

Colin-Seifer commented on docs/document.md:
```diff
@@ -10,6 +10,8 @@ The `Django` subclass contains parameters related to Django's side of the docume
 * `queryset_pagination` (*optional*) - Size of the chunk when indexing,
   override [`OPENSEARCH_DSL_QUERYSET_PAGINATION`](settings.md#opensearch_dsl_queryset_pagination.md).
+* `related_models` (*optional*) - Specifies a relation between models that allows for index updating based on these
+  defined relationships.
```
Reworded this in my latest update

Colin-Seifer commented on django_opensearch_dsl/documents.py:
```diff
@@ -132,8 +135,15 @@ def init_prepare(self):
 
-            prep_func = getattr(self, "prepare_%s" % name, None)
-            fn = prep_func if prep_func else partial(field.get_value_from_instance)
+            prep_func = getattr(self, "prepare_%s_with_related" % name, None)
+            if prep_func:
+                fn = partial(prep_func, related_to_ignore=self._related_instance_to_ignore)
+            else:
+                prep_func = getattr(self, "prepare_%s" % name, None)
+                if prep_func:
+                    fn = prep_func
+                else:
+                    fn = partial(field.get_value_from_instance, field_value_to_ignore=self._related_instance_to_ignore)
```
The fields documentation wasn't the appropriate place to talk about related_instance_to_ignore, but I did add quite a bit to the docstring in the `delete_related` method in `registries.py`. Basically related_instance_to_ignore makes sure the index is appropriately updated and objects aren't orphaned on a delete. I also added documentation on `prepare_%s_with_related`. From what I gather from looking over the elasticsearch history, it's a somewhat uncommon use case, but basically just makes sure that the index updates properly.

qcoumes commented:
Thanks for your PR !

I'll try to make a release with this feature and #15 tomorrow.
